pywren/scripts/standalone.py

```diff
@@ -310,41 +310,57 @@ def server(aws_region, max_run_time, run_dir, sqs_queue_name, max_idle_time,
-                     .format(server_name)
-
-    formatter = logging.Formatter(log_format_str, "%Y-%m-%d %H:%M:%S")
-
-
-    handler = watchtower.CloudWatchLogHandler(send_interval=20,
-                                              log_group="pywren.standalone",
-                                              stream_name=log_stream_prefix + "-{logger_name}",
-                                              boto3_session=session,
-                                              max_batch_count=10)
-
-    debug_stream_handler = logging.StreamHandler()
-    handler.setFormatter(formatter)
-    logger.addHandler(handler)
-    logger.setLevel(logging.DEBUG)
-
-    wren_log = pywren.wrenhandler.logger
-    wren_log.addHandler(handler)
-
-    wren_log.addHandler(debug_stream_handler)
+    def async_log_setup():
+        ''' None of this stuff should be on the critical path to launching an
+            * instance. Instances should start dequeuing from SQS queue as soon
+            * as possible and shouldn't have to wait for rest of spot cluster
+            * to come up so they have a valid ec2_metadata['Name']
+            * If there are any exceptions in this function,
+            * we should exponentially backoff and try again until we succeed,
+            * this is critical because if this doesn't happen we end up
+            * clogging all EC2 resources
+        '''
+        success = False
+        backoff_time = 5
+        while (not success):
+            try:
+                time.sleep(backoff_time)
+                instance = get_my_ec2_instance(aws_region)
+                ec2_metadata = get_my_ec2_meta(instance)
+                server_name = ec2_metadata['Name']
+                log_stream_prefix = ec2_metadata['instance_id']
+
+                log_format_str = '{} %(asctime)s - %(name)s- %(levelname)s - %(message)s'\
+                                 .format(server_name)
+
+                formatter = logging.Formatter(log_format_str, "%Y-%m-%d %H:%M:%S")
+                stream_name = log_stream_prefix + "-{logger_name}"
+                handler = watchtower.CloudWatchLogHandler(send_interval=20,
+                                                          log_group="pywren.standalone",
+                                                          stream_name=stream_name,
+                                                          boto3_session=session,
+                                                          max_batch_count=10)
+
+                debug_stream_handler = logging.StreamHandler()
+                handler.setFormatter(formatter)
+                logger.addHandler(handler)
+                logger.setLevel(logging.DEBUG)
+                wren_log = pywren.wrenhandler.logger
+                wren_log.addHandler(handler)
+                wren_log.addHandler(debug_stream_handler)
+                success = True
+            except:
+                backoff_time *= 2
+
```

Comment:
So one thing that will be very good to do is demarcate what will be called once and what will be called once per-job dequeued from SQS. My understanding is that server is called once per machine -- is that right ?

Reply:
once per "process"

Reply:
Can you name this constant something like "EXP_BACKOFF_FACTOR" or something? I just hate having these mystery constants hanging out